# Analysis

### Aim:
   * Aim is to find the ranges of all hyperparameters for every model which are generated with different runtimes.
   * Finding the important hyperparameters.
   * Comparing the hyperparameters range
          

In [205]:
import numpy as np
import pandas as pd

### Analysis for GBM

**Loading the csv of GBM**

In [206]:
df_gbm=pd.read_csv("gbm_actual.csv", decimal = ',')

In [207]:
df_gbm.head()

Unnamed: 0                                    modelId learn_rate  \
0           0  GBM_grid_1_AutoML_20190415_223833_model_4      0.008   
1           1               GBM_4_AutoML_20190415_223833        0.1   
2           2               GBM_3_AutoML_20190415_223833        0.1   
3           3  GBM_grid_1_AutoML_20190415_223833_model_7      0.005   
4           4               GBM_2_AutoML_20190415_223833        0.1   

  learn_rate_annealing col_sample_rate    max_abs_leafnode_pred  \
0                  1.0             0.7  1.7976931348623157e+308   
1                  1.0             0.8  1.7976931348623157e+308   
2                  1.0             0.8  1.7976931348623157e+308   
3                  1.0             0.7  1.7976931348623157e+308   
4                  1.0             0.8  1.7976931348623157e+308   

  pred_noise_bandwidth distribution tweedie quantile_alpha huber_alpha  \
0                  0.0    bernoulli     1.5            0.5         0.9   
1                  0.0    bernoulli     1.5            0.5         0.9   
2                  0.0    bernoulli     1.5            0.5         0.9   
3                  0.0    bernoulli     1.5            0.5         0.9   
4                  0.0    bernoulli     1.5            0.5         0.9   

  categorical_encoding histogram_type  ntrees  max_depth min_rows sample_rate  \
0                 AUTO           AUTO     295         16     30.0         0.7   
1                 AUTO           AUTO      96         15    100.0         0.8   
2                 AUTO           AUTO      96         15    100.0         0.8   
3                 AUTO           AUTO     165         13     15.0         0.6   
4                 AUTO           AUTO      96         15    100.0         0.8   

  col_sample_rate_per_tree min_split_improvement  
0                      0.7                 1e-05  
1                      0.8                 1e-05  
2                      0.8                 1e-05  
3                      0.7                 1e-05  
4                      0.8                 1e-05

In [208]:
df_gbm.drop(df_gbm.columns[0] , axis=1)

modelId learn_rate  \
0    GBM_grid_1_AutoML_20190415_223833_model_4      0.008   
1                 GBM_4_AutoML_20190415_223833        0.1   
2                 GBM_3_AutoML_20190415_223833        0.1   
3    GBM_grid_1_AutoML_20190415_223833_model_7      0.005   
4                 GBM_2_AutoML_20190415_223833        0.1   
5                 GBM_5_AutoML_20190415_223833        0.1   
6    GBM_grid_1_AutoML_20190415_223833_model_2       0.05   
7                 GBM_1_AutoML_20190415_223833        0.1   
8    GBM_grid_1_AutoML_20190415_223833_model_1      0.001   
9    GBM_grid_1_AutoML_20190415_223833_model_3       0.01   
10   GBM_grid_1_AutoML_20190415_223833_model_8        0.1   
11   GBM_grid_1_AutoML_20190415_223833_model_6      0.008   
12   GBM_grid_1_AutoML_20190415_223833_model_5      0.005   
13   GBM_grid_1_AutoML_20190418_130009_model_1       0.01   
14                GBM_4_AutoML_20190418_130009        0.1   
15                GBM_3_AutoML_20190418_130009        0.1   
16                GBM_5_AutoML_20190418_130009        0.1   
17                GBM_2_AutoML_20190418_130009        0.1   
18                GBM_1_AutoML_20190418_130009        0.1   
19   GBM_grid_1_AutoML_20190418_130009_model_6        0.1   
20   GBM_grid_1_AutoML_20190418_130009_model_2        0.8   
21   GBM_grid_1_AutoML_20190418_130009_model_5        0.5   
22   GBM_grid_1_AutoML_20190418_130009_model_8       0.08   
23   GBM_grid_1_AutoML_20190418_130009_model_3       0.01   
24   GBM_grid_1_AutoML_20190418_130009_model_4      0.001   
25   GBM_grid_1_AutoML_20190418_130009_model_7      0.001   
26   GBM_grid_1_AutoML_20190418_120619_model_2        0.1   
27   GBM_grid_1_AutoML_20190418_120619_model_5       0.05   
28   GBM_grid_1_AutoML_20190418_120619_model_9       0.01   
29                GBM_4_AutoML_20190418_120619        0.1   
30                GBM_3_AutoML_20190418_120619        0.1   
31                GBM_5_AutoML_20190418_120619        0.1   
32                GBM_2_AutoML_20190418_120619        0.1   
33                GBM_1_AutoML_20190418_120619        0.1   
34  GBM_grid_1_AutoML_20190418_120619_model_15        0.1   
35  GBM_grid_1_AutoML_20190418_120619_model_14       0.01   
36   GBM_grid_1_AutoML_20190418_120619_model_6       0.01   
37   GBM_grid_1_AutoML_20190418_120619_model_7        0.8   
38  GBM_grid_1_AutoML_20190418_120619_model_16       0.08   
39  GBM_grid_1_AutoML_20190418_120619_model_11        0.1   
40   GBM_grid_1_AutoML_20190418_120619_model_8        0.5   
41   GBM_grid_1_AutoML_20190418_120619_model_1        0.1   
42   GBM_grid_1_AutoML_20190418_120619_model_3        0.5   
43  GBM_grid_1_AutoML_20190418_120619_model_12        0.8   
44   GBM_grid_1_AutoML_20190418_120619_model_4      0.005   
45  GBM_grid_1_AutoML_20190418_120619_model_10       0.01   
46  GBM_grid_1_AutoML_20190418_120619_model_13      0.005   

   learn_rate_annealing col_sample_rate    max_abs_leafnode_pred  \
0                   1.0             0.7  1.7976931348623157e+308   
1                   1.0             0.8  1.7976931348623157e+308   
2                   1.0             0.8  1.7976931348623157e+308   
3                   1.0             0.7  1.7976931348623157e+308   
4                   1.0             0.8  1.7976931348623157e+308   
5                   1.0             0.8  1.7976931348623157e+308   
6                   1.0             1.0  1.7976931348623157e+308   
7                   1.0             0.8  1.7976931348623157e+308   
8                   1.0             1.0  1.7976931348623157e+308   
9                   1.0             0.7  1.7976931348623157e+308   
10                  1.0             0.4  1.7976931348623157e+308   
11                  1.0             0.7  1.7976931348623157e+308   
12                  1.0             0.7  1.7976931348623157e+308   
13                  1.0             0.7  1.7976931348623157e+308   
14                  1.0             0.8  1.7976931348623157e+308   
15                  1.

In [209]:
len(df_gbm)

47

In [210]:
df_gbm.dtypes

Unnamed: 0                   int64
modelId                     object
learn_rate                  object
learn_rate_annealing        object
col_sample_rate             object
max_abs_leafnode_pred       object
pred_noise_bandwidth        object
distribution                object
tweedie                     object
quantile_alpha              object
huber_alpha                 object
categorical_encoding        object
histogram_type              object
ntrees                       int64
max_depth                    int64
min_rows                    object
sample_rate                 object
col_sample_rate_per_tree    object
min_split_improvement       object
dtype: object

**Dropping the colums with the values which are not responsible to give any range** 

In [211]:
df_gbm['distribution'].head()

0    bernoulli
1    bernoulli
2    bernoulli
3    bernoulli
4    bernoulli
Name: distribution, dtype: object

In [212]:
df_gbm['categorical_encoding'].head()

0    AUTO
1    AUTO
2    AUTO
3    AUTO
4    AUTO
Name: categorical_encoding, dtype: object

In [213]:
df_gbm['histogram_type'].head()

0    AUTO
1    AUTO
2    AUTO
3    AUTO
4    AUTO
Name: histogram_type, dtype: object

**Dropping hyperparameter max_abs_leafnode_pred as it is giving the constant value throughout that is 1.7976931348623157e+308 so the range will also be the same**

In [214]:
df_gbm['max_abs_leafnode_pred'].head()

0    1.7976931348623157e+308
1    1.7976931348623157e+308
2    1.7976931348623157e+308
3    1.7976931348623157e+308
4    1.7976931348623157e+308
Name: max_abs_leafnode_pred, dtype: object

In [215]:
df_gbm= df_gbm.drop(columns=['distribution','categorical_encoding','histogram_type','max_abs_leafnode_pred'])
df_gbm

Unnamed: 0                                     modelId learn_rate  \
0            0   GBM_grid_1_AutoML_20190415_223833_model_4      0.008   
1            1                GBM_4_AutoML_20190415_223833        0.1   
2            2                GBM_3_AutoML_20190415_223833        0.1   
3            3   GBM_grid_1_AutoML_20190415_223833_model_7      0.005   
4            4                GBM_2_AutoML_20190415_223833        0.1   
5            5                GBM_5_AutoML_20190415_223833        0.1   
6            6   GBM_grid_1_AutoML_20190415_223833_model_2       0.05   
7            7                GBM_1_AutoML_20190415_223833        0.1   
8            8   GBM_grid_1_AutoML_20190415_223833_model_1      0.001   
9            9   GBM_grid_1_AutoML_20190415_223833_model_3       0.01   
10          10   GBM_grid_1_AutoML_20190415_223833_model_8        0.1   
11          11   GBM_grid_1_AutoML_20190415_223833_model_6      0.008   
12          12   GBM_grid_1_AutoML_20190415_223833_model_5      0.005   
13          13   GBM_grid_1_AutoML_20190418_130009_model_1       0.01   
14          14                GBM_4_AutoML_20190418_130009        0.1   
15          15                GBM_3_AutoML_20190418_130009        0.1   
16          16                GBM_5_AutoML_20190418_130009        0.1   
17          17                GBM_2_AutoML_20190418_130009        0.1   
18          18                GBM_1_AutoML_20190418_130009        0.1   
19          19   GBM_grid_1_AutoML_20190418_130009_model_6        0.1   
20          20   GBM_grid_1_AutoML_20190418_130009_model_2        0.8   
21          21   GBM_grid_1_AutoML_20190418_130009_model_5        0.5   
22          22   GBM_grid_1_AutoML_20190418_130009_model_8       0.08   
23          23   GBM_grid_1_AutoML_20190418_130009_model_3       0.01   
24          24   GBM_grid_1_AutoML_20190418_130009_model_4      0.001   
25          25   GBM_grid_1_AutoML_20190418_130009_model_7      0.001   
26          26   GBM_grid_1_AutoML_20190418_120619_model_2        0.1   
27          27   GBM_grid_1_AutoML_20190418_120619_model_5       0.05   
28          28   GBM_grid_1_AutoML_20190418_120619_model_9       0.01   
29          29                GBM_4_AutoML_20190418_120619        0.1   
30          30                GBM_3_AutoML_20190418_120619        0.1   
31          31                GBM_5_AutoML_20190418_120619        0.1   
32          32                GBM_2_AutoML_20190418_120619        0.1   
33          33                GBM_1_AutoML_20190418_120619        0.1   
34          34  GBM_grid_1_AutoML_20190418_120619_model_15        0.1   
35          35  GBM_grid_1_AutoML_20190418_120619_model_14       0.01   
36          36   GBM_grid_1_AutoML_20190418_120619_model_6       0.01   
37          37   GBM_grid_1_AutoML_20190418_120619_model_7        0.8   
38          38  GBM_grid_1_AutoML_20190418_120619_model_16       0.08   
39          39  GBM_grid_1_AutoML_20190418_120619_model_11        0.1   
40          40   GBM_grid_1_AutoML_20190418_120619_model_8        0.5   
41          41   GBM_grid_1_AutoML_20190418_120619_model_1        0.1   
42          42   GBM_grid_1_AutoML_20190418_120619_model_3        0.5   
43          43  GBM_grid_1_AutoML_20190418_120619_model_12        0.8   
44          44   GBM_grid_1_AutoML_20190418_120619_model_4      0.005   
45          45  GBM_grid_1_AutoML_20190418_120619_model_10       0.01   
46          46  GBM_grid_1_AutoML_20190418_120619_model_13      0.005   

   learn_rate_annealing col_sample_rate pred_noise_bandwidth tweedie  \
0                   1.0             0.7                  0.0     1.5   
1                   1.0             0.8                  0.0     1.5   
2                   1.0             0.8                  0.0     1.5   
3                   1.0             0.7                  0.0     1.5   
4                   1.0             0.8                  0.0     1.5   
5                   1.0             0.8                  0.0     1.

In [216]:
a = df_gbm.iloc[:,2:]
a

learn_rate learn_rate_annealing col_sample_rate pred_noise_bandwidth  \
0       0.008                  1.0             0.7                  0.0   
1         0.1                  1.0             0.8                  0.0   
2         0.1                  1.0             0.8                  0.0   
3       0.005                  1.0             0.7                  0.0   
4         0.1                  1.0             0.8                  0.0   
5         0.1                  1.0             0.8                  0.0   
6        0.05                  1.0             1.0                  0.0   
7         0.1                  1.0             0.8                  0.0   
8       0.001                  1.0             1.0                  0.0   
9        0.01                  1.0             0.7                  0.0   
10        0.1                  1.0             0.4                  0.0   
11      0.008                  1.0             0.7                  0.0   
12      0.005                  1.0             0.7                  0.0   
13       0.01                  1.0             0.7                  0.0   
14        0.1                  1.0             0.8                  0.0   
15        0.1                  1.0             0.8                  0.0   
16        0.1                  1.0             0.8                  0.0   
17        0.1                  1.0             0.8                  0.0   
18        0.1                  1.0             0.8                  0.0   
19        0.1                  1.0             0.4                  0.0   
20        0.8                  1.0             0.7                  0.0   
21        0.5                  1.0             1.0                  0.0   
22       0.08                  1.0             0.7                  0.0   
23       0.01                  1.0             0.7                  0.0   
24      0.001                  1.0             0.7                  0.0   
25      0.001                  1.0             0.7                  0.0   
26        0.1                  1.0             1.0                  0.0   
27       0.05                  1.0             0.4                  0.0   
28       0.01                  1.0             0.7                  0.0   
29        0.1                  1.0             0.8                  0.0   
30        0.1                  1.0             0.8                  0.0   
31        0.1                  1.0             0.8                  0.0   
32        0.1                  1.0             0.8                  0.0   
33        0.1                  1.0             0.8                  0.0   
34        0.1                  1.0             0.4                  0.0   
35       0.01                  1.0             0.4                  0.0   
36       0.01                  1.0             0.7                  0.0   
37        0.8                  1.0             1.0                  0.0   
38       0.08                  1.0             1.0                  0.0   
39        0.1                  1.0             0.7                  0.0   
40        0.5                  1.0             0.4                  0.0   
41        0.1                  1.0             0.4                  0.0   
42        0.5                  1.0             1.0                  0.0   
43        0.8                  1.0             0.7                  0.0   
44      0.005                  1.0             0.4                  0.0   
45       0.01                  1.0             0.4                  0.0   
46      0.005                  1.0             0.7                  0.0   

   tweedie quantile_alpha huber_alpha  ntrees  max_depth min_rows sample_rate  \
0      1.5            0.5         0.9     295         16     30.0         0.7   
1      1.5            0.5         0.9      96         15    100.0         0.8   
2      1.5            0.5         0.9      96         15    100.0         0.8   
3      1.5            0.5         0.9     165         13     15.0         0.6 

**Converting the object data to numeric data type**

In [217]:
df_gbm = df_gbm.iloc[:,2:].apply(pd.to_numeric, downcast='integer')

In [218]:
df_gbm.dtypes

learn_rate                  float64
learn_rate_annealing           int8
col_sample_rate             float64
pred_noise_bandwidth           int8
tweedie                     float64
quantile_alpha              float64
huber_alpha                 float64
ntrees                        int16
max_depth                      int8
min_rows                       int8
sample_rate                 float64
col_sample_rate_per_tree    float64
min_split_improvement       float64
dtype: object

In [219]:
ranges = []
for key,val in df_gbm.dtypes.items():
    print(key,val)
    if val == 'int': 
         ranges.append([key,[df_gbm[key].min(),df_gbm[key].max()]])
    if val == 'float':
         ranges.append([key,[df_gbm[key].min(),df_gbm[key].max()]])

learn_rate float64
learn_rate_annealing int8
col_sample_rate float64
pred_noise_bandwidth int8
tweedie float64
quantile_alpha float64
huber_alpha float64
ntrees int16
max_depth int8
min_rows int8
sample_rate float64
col_sample_rate_per_tree float64
min_split_improvement float64


### Finding the ranges of the hyper parameter

In [220]:
ranges

[['learn_rate', [0.001, 0.8]],
 ['col_sample_rate', [0.4, 1.0]],
 ['tweedie', [1.5, 1.5]],
 ['quantile_alpha', [0.5, 0.5]],
 ['huber_alpha', [0.9, 0.9]],
 ['sample_rate', [0.5, 1.0]],
 ['col_sample_rate_per_tree', [0.4, 1.0]],
 ['min_split_improvement', [1e-05, 0.0001]]]

## Range of hyperparameters for GBM

|Hyperparameter| Minimum | Maximum |
|---|---|---|
|learn_rate|0.001|0.8|
|col_sample_rate|0.4|1.0|
|tweedie|1.5|1.5|
|quantile_alpha|0.5| 0.5|
|huber_alpha|0.9| 0.9|
|sample_rate|0.5| 1.0|
|col_sample_rate_per_tree|0.4| 1.0|
|min_split_improvement|1e-05| 0.0001|
|max_abs_leafnode_pred|1.7976931348623157e+308|1.7976931348623157e+308|

### Analysis for DRF

**Loading the csv of DRF**

In [221]:
df_drf=pd.read_csv("drf_actual.csv", decimal = ',')

In [222]:
df_drf.head()

Unnamed: 0                       modelId  mtries categorical_encoding
0           0  DRF_1_AutoML_20190415_223833      -1                 AUTO
1           1  DRF_1_AutoML_20190418_113508      -1                 AUTO
2           2  DRF_1_AutoML_20190418_130009      -1                 AUTO
3           3  DRF_1_AutoML_20190418_120619      -1                 AUTO

In [223]:
len(df_drf)

4

In [224]:
df_drf.dtypes

Unnamed: 0               int64
modelId                 object
mtries                   int64
categorical_encoding    object
dtype: object

In [225]:
range_mtries = [df_drf['mtries'].min(),df_drf['mtries'].max()]
range_mtries

[-1, -1]

## Range of hyperparameters for DRF

|Hyperparameter| Minimum | Maximum |
|---|---|---|
|mtries|-1 | -1|

### Analysis for XRT

**Loading the csv of XRT**

In [226]:
df_xrt=pd.read_csv("xrt_actual.csv", decimal = ',')

In [227]:
df_xrt.head()

Unnamed: 0                       modelId  mtries categorical_encoding
0           0  XRT_1_AutoML_20190415_223833      -1                 AUTO
1           1  XRT_1_AutoML_20190418_113508      -1                 AUTO
2           2  XRT_1_AutoML_20190418_130009      -1                 AUTO
3           3  XRT_1_AutoML_20190418_120619      -1                 AUTO

In [228]:
len(df_xrt)

4

In [229]:
df_xrt.dtypes

Unnamed: 0               int64
modelId                 object
mtries                   int64
categorical_encoding    object
dtype: object

In [230]:
range_mtries = [df_xrt['mtries'].min(),df_xrt['mtries'].max()]
range_mtries

[-1, -1]

## Range of hyperparameters for XRT

|Hyperparameter| Minimum | Maximum |
|---|---|---|
|mtries|-1 | -1|

### Analysis for GLM

**Loading the csv of GLM**

In [231]:
df_glm=pd.read_csv("glm_actual.csv", decimal = ',')

In [232]:
df_glm.head()

Unnamed: 0                                    modelId                 seed  \
0           0  GLM_grid_1_AutoML_20190415_223833_model_1   888573333419439396   
1           1  GLM_grid_1_AutoML_20190418_113508_model_1   711766579752531550   
2           2  GLM_grid_1_AutoML_20190418_130009_model_1  7150684335027209933   
3           3  GLM_grid_1_AutoML_20190418_120619_model_1  6743050173023044500   

  tweedie_variance_power tweedie_link_power                           alpha  \
0                    0.0                1.0  [0.0, 0.2, 0.4, 0.6, 0.8, 1.0]   
1                    0.0                1.0  [0.0, 0.2, 0.4, 0.6, 0.8, 1.0]   
2                    0.0                1.0  [0.0, 0.2, 0.4, 0.6, 0.8, 1.0]   
3                    0.0                1.0  [0.0, 0.2, 0.4, 0.6, 0.8, 1.0]   

                                                                  lambda  \
0  [0.40044436071981493, 0.24868273229194082, 0.15443618990917896, 0....   
1  [0.40044436071981493, 0.24868273229194082, 0.15443618990917896, 0....   
2  [0.40044436071981493, 0.24868273229194082, 0.15443618990917896, 0....   
3  [0.40044436071981493, 0.24868273229194082, 0.15443618990917896, 0....   

  missing_values_handling  standardize         theta  
0          MeanImputation         True  1.000000e-10  
1          MeanImputation         True           NaN  
2          MeanImputation         True  1.000000e-10  
3          MeanImputation         True  1.000000e-10

[0.0, 0.2, 0.4, 0.6, 0.8, 1.0] = alpha

In [233]:
alpha_list = ['0.0', '0.2', '0.4', '0.6', '0.8', '1.0']

In [234]:
alpha = [float(i) for i in alpha_list] 
alpha

[0.0, 0.2, 0.4, 0.6, 0.8, 1.0]

In [235]:
range_mtries = [min(alpha),max(alpha)]
range_mtries

[0.0, 1.0]

In [236]:
lambda_list = ['0.40044436071981493', '0.24868273229194082', '0.15443618990917896', '0.09590749037478274', '0.05956017637704195', '0.036987878592192616', '0.022970099250379815', '0.014264820791416522', '0.008858695384516653', '0.005501399916841423', '0.003416462552479351', '0.0021216811264277652', '0.0013175999248032143']

In [237]:
lambda_list = [float(i) for i in lambda_list] 
lambda_list

[0.40044436071981493,
 0.24868273229194082,
 0.15443618990917896,
 0.09590749037478274,
 0.05956017637704195,
 0.036987878592192616,
 0.022970099250379815,
 0.014264820791416522,
 0.008858695384516653,
 0.005501399916841423,
 0.003416462552479351,
 0.0021216811264277652,
 0.0013175999248032143]

In [238]:
range_mtries = [min(lambda_list),max(lambda_list)]
range_mtries

[0.0013175999248032143, 0.40044436071981493]

In [239]:
range_mtries = [df_glm['seed'].min(),df_glm['seed'].max()]
range_mtries

[711766579752531550, 7150684335027209933]

In [240]:
range_mtries = [df_glm['tweedie_variance_power'].min(),df_glm['tweedie_variance_power'].max()]
range_mtries

['0.0', '0.0']

In [241]:
range_mtries = [df_glm['tweedie_link_power'].min(),df_glm['tweedie_link_power'].max()]
range_mtries

['1.0', '1.0']

In [242]:
range_mtries = [df_glm['theta'].min(),df_glm['theta'].max()]
range_mtries

[1e-10, 1e-10]

## Range of hyperparameters for GLM

|Hyperparameter| Minimum | Maximum |
|---|---|---|
|seed|711766579752531550 | 7150684335027209933|
|tweedie_variance_power|0.0 | 0.0|
|tweedie_link_power|1.0|1.0|
|alpha|0.0 | 1.0|
|lambda|0.0013175999248032143| 0.40044436071981493|
|theta|1e-10| 1e-10|

## Comparing Hyperparameters

* According to the analysis it is found that **DRF** and **XRF** are having the same hyperparameter which is **mtries** 

* For DRF and XRT mtries have the same value of **-1** throughout, that means its constant and not changing.

|Model|Hyperparameter|range|
|---|---|---|
|DRF| mtries | -1|
|XRT| mtries | -1|

# Finding Important Hyperparameters

Important hyperparameters are those specific hyperparameters which on different values in their given range affects the performance of the model.  

In [243]:
from sklearn.ensemble import RandomForestClassifier,AdaBoostClassifier

In [244]:
df=pd.read_csv("Employee_access_data.csv", decimal = ',')

In [245]:
X = df[['RESOURCE','MGR_ID','ROLE_ROLLUP_1','ROLE_ROLLUP_2','ROLE_DEPTNAME','ROLE_TITLE','ROLE_FAMILY_DESC','ROLE_FAMILY','ROLE_CODE']]
Y = df['ACTION']

In [246]:
# grid_search.fit(X,Y)

In [247]:
from sklearn.tree import DecisionTreeClassifier
AdaBoostClassifier().get_params().keys()

dict_keys(['algorithm', 'base_estimator', 'learning_rate', 'n_estimators', 'random_state'])

In [248]:
RandomForestClassifier().get_params().keys()

dict_keys(['bootstrap', 'class_weight', 'criterion', 'max_depth', 'max_features', 'max_leaf_nodes', 'min_impurity_decrease', 'min_impurity_split', 'min_samples_leaf', 'min_samples_split', 'min_weight_fraction_leaf', 'n_estimators', 'n_jobs', 'oob_score', 'random_state', 'verbose', 'warm_start'])

In [249]:
AdaBoostClassifier().get_params().keys()

dict_keys(['algorithm', 'base_estimator', 'learning_rate', 'n_estimators', 'random_state'])

In [250]:
import xgboost as xgb
from xgboost import XGBClassifier
XGBClassifier().get_params().keys()

dict_keys(['base_score', 'booster', 'colsample_bylevel', 'colsample_bytree', 'gamma', 'learning_rate', 'max_delta_step', 'max_depth', 'min_child_weight', 'missing', 'n_estimators', 'n_jobs', 'nthread', 'objective', 'random_state', 'reg_alpha', 'reg_lambda', 'scale_pos_weight', 'seed', 'silent', 'subsample'])

## Grid Search to Find important hyperparameters


Grid search is the process of performing hyper parameter tuning in order to determine the optimal values for a given model. This is significant as the performance of the entire model is based on the hyper parameter values specified. It works by searching exhaustively through a specified subset of hyperparameters. The benefit of grid search is that it is guaranteed to find the optimal combination of parameters supplied. The drawback is that it can be very time consuming and computationally expensive. 


### H2O

In [251]:
#importing h2o libraries
import h2o
from h2o.automl import H2OAutoML
import random, os, sys
from datetime import datetime
import pandas as pd
import logging
import csv
import optparse
import time
import json
from distutils.util import strtobool
import psutil
import numpy as np

In [252]:
port_no=random.randint(5555,55555)
h2o.init(strict_version_check=False,port=port_no)
#importing train file

df = h2o.import_file(path = 'Employee_access_data.csv')

Checking whether there is an H2O instance running at http://localhost:6794 ..... not found.
Attempting to start a local H2O server...
; Java HotSpot(TM) 64-Bit Server VM 18.9 (build 11.0.2+9-LTS, mixed mode)
  Starting server from C:\Users\hp\Anaconda3\lib\site-packages\h2o\backend\bin\h2o.jar
  Ice root: C:\Users\hp\AppData\Local\Temp\tmp0azj8we8
  JVM stdout: C:\Users\hp\AppData\Local\Temp\tmp0azj8we8\h2o_hp_started_from_python.out
  JVM stderr: C:\Users\hp\AppData\Local\Temp\tmp0azj8we8\h2o_hp_started_from_python.err
  Server is running at http://127.0.0.1:6794
Connecting to H2O server at http://127.0.0.1:6794 ... successful.


--------------------------  ------------------------------------------
H2O cluster uptime:         07 secs
H2O cluster timezone:       America/New_York
H2O data parsing timezone:  UTC
H2O cluster version:        3.24.0.1
H2O cluster version age:    25 days
H2O cluster name:           H2O_from_python_hp_a7lzof
H2O cluster total nodes:    1
H2O cluster free memory:    1.984 Gb
H2O cluster total cores:    4
H2O cluster allowed cores:  4
H2O cluster status:         accepting new members, healthy
H2O connection url:         http://127.0.0.1:6794
H2O connection proxy:
H2O internal security:      False
H2O API Extensions:         Amazon S3, Algos, AutoML, Core V3, Core V4
Python version:             3.7.1 final
--------------------------  ------------------------------------------

Parse progress: |█████████████████████████████████████████████████████████| 100%


In [253]:
def get_independent_variables(df, targ):
   C = [name for name in df.columns if name != targ]
   # determine column types
   ints, reals, enums = [], [], []
   for key, val in df.types.items():
       if key in C:
           if val == 'enum':
               enums.append(key)
           elif val == 'int':
               ints.append(key)
           else:
               reals.append(key)
   x=ints+enums+reals
   return x

In [254]:
X=get_independent_variables(df, 'ACTION')
print(X)
y = 'ACTION'

['RESOURCE', 'MGR_ID', 'ROLE_ROLLUP_1', 'ROLE_ROLLUP_2', 'ROLE_DEPTNAME', 'ROLE_TITLE', 'ROLE_FAMILY_DESC', 'ROLE_FAMILY', 'ROLE_CODE']


### Grid search for DRF

In [255]:
drf_params1 = { 'mtries': [-2,-1,1,2,3] }

Passing -3 in the range of mtries will give error it must be -1 or -2 or >=1

In [256]:
from h2o.estimators import H2ORandomForestEstimator
from h2o.grid.grid_search import H2OGridSearch
# Train and validate a cartesian grid of GBMs
drf_grid1 = H2OGridSearch(model=H2ORandomForestEstimator,
                          grid_id='drf',
                          hyper_params=drf_params1)

In [257]:
drf_grid1.train(X,y, training_frame=df)

drf Grid Build progress: |████████████████████████████████████████████████| 100%


In [258]:
drf_grid1 = drf_grid1.get_grid(sort_by='r2')
drf_grid1

    mtries    model_ids                   r2
0        1  drf_model_3  0.09137774408807087
1        2  drf_model_4  0.22582166574132467
2       -2  drf_model_1  0.23213009571304521
3        3  drf_model_5   0.2522982863593466
4       -1  drf_model_2   0.2571045224125067


In [259]:
type(drf_grid1)

h2o.grid.grid_search.H2OGridSearch

### Grid Search for GLM

In [260]:
glm_params1 = { 'tweedie_variance_power': [-1,0,1]}

In [261]:
from h2o.estimators import H2ORandomForestEstimator
from h2o.estimators import H2OGeneralizedLinearEstimator
from h2o.grid.grid_search import H2OGridSearch
# Train and validate a cartesian grid of GBMs
glm_grid1 = H2OGridSearch(model=H2OGeneralizedLinearEstimator,
                          grid_id='glm',
                          hyper_params=glm_params1)

In [262]:
glm_grid1.train(X,y, training_frame=df)

glm Grid Build progress: |████████████████████████████████████████████████| 100%


In [263]:
glm_grid1 = glm_grid1.get_grid(sort_by='rmse')
glm_grid1

    tweedie_variance_power    model_ids                 rmse
0                      0.0  glm_model_2  0.23345202416059152
1                      1.0  glm_model_3  0.23345202416059152
2                     -1.0  glm_model_1  0.23345202416059152


In [264]:
glm_params2 = {'tweedie_variance_power': [-1,0,1],'tweedie_link_power': [0,1,2]
#                 'alpha' : [0,0.2,0.4,0.7,0.8,1],
#                 'lambda': [0.0013175999248032143,0.09590749037478274, 0.022970099250379815, 0.003416462552479351,0.40044436071981493],
#                 'theta': [888573333419439000,711766579752531000,6743050173023040000]
              }

In [265]:
from h2o.estimators import H2ORandomForestEstimator
from h2o.estimators import H2OGeneralizedLinearEstimator
from h2o.grid.grid_search import H2OGridSearch
# Train and validate a cartesian grid of GBMs
glm_grid2 = H2OGridSearch(model=H2OGeneralizedLinearEstimator,
                          grid_id='glm',
                          hyper_params=glm_params2)

In [266]:
glm_grid2.train(X,y, training_frame=df)

glm Grid Build progress: |████████████████████████████████████████████████| 100%


In [267]:
glm_grid2 = glm_grid2.get_grid(sort_by='rmse')
glm_grid2

    tweedie_variance_power     model_ids                 rmse
0                      0.0   glm_model_7  0.23345202416059152
1                      0.0   glm_model_2  0.23345202416059152
2                      1.0   glm_model_3  0.23345202416059152
3                     -1.0   glm_model_6  0.23345202416059152
4                     -1.0   glm_model_1  0.23345202416059152
5                     -1.0   glm_model_4  0.23345202416059152
6                      1.0  glm_model_10  0.23345202416059152
7                      1.0  glm_model_12  0.23345202416059152
8                      0.0   glm_model_9  0.23345202416059152


**tweedie_link_power is not affecting the score thus we can not consider it as important

In [268]:
glm_params1 = {'tweedie_link_power': [-1,0,1,2]
#                 'alpha' : [0,0.2,0.4,0.7,0.8,1],
#                 'lambda': [0.0013175999248032143,0.09590749037478274, 0.022970099250379815, 0.003416462552479351,0.40044436071981493],
#                 'theta': [888573333419439000,711766579752531000,6743050173023040000]
              }
from h2o.estimators import H2ORandomForestEstimator
from h2o.estimators import H2OGeneralizedLinearEstimator
from h2o.grid.grid_search import H2OGridSearch
# Train and validate a cartesian grid of GBMs
glm_grid1 = H2OGridSearch(model=H2OGeneralizedLinearEstimator,
                          grid_id='glm',
                          hyper_params=glm_params1)

In [269]:
glm_grid1.train(X,y, training_frame=df)

glm Grid Build progress: |████████████████████████████████████████████████| 100%


In [270]:
glm_grid1 = glm_grid1.get_grid(sort_by='rmse')
glm_grid1

    tweedie_variance_power     model_ids                 rmse
0                      0.0   glm_model_7  0.23345202416059152
1                      0.0   glm_model_2  0.23345202416059152
2                      1.0   glm_model_3  0.23345202416059152
3                     -1.0   glm_model_6  0.23345202416059152
4                     -1.0   glm_model_1  0.23345202416059152
5                     -1.0   glm_model_4  0.23345202416059152
6                      0.0  glm_model_10  0.23345202416059152
7                      1.0  glm_model_12  0.23345202416059152
8                      0.0  glm_model_10  0.23345202416059152
9                      0.0   glm_model_9  0.23345202416059152


**tweedie_link_power can not be considered as important hyperparameterr because score is not getting affected**

In [271]:
glm_params1 = { 'tweedie_variance_power': [-1,0,1],
                 'alpha' : [0,0.2,0.7,1]
#                 'lambda': [0.0013175999248032143,0.09590749037478274, 0.022970099250379815, 0.003416462552479351,0.40044436071981493],
#                 'theta': [888573333419439000,711766579752531000,6743050173023040000]
              }

In [272]:
from h2o.estimators import H2ORandomForestEstimator
from h2o.estimators import H2OGeneralizedLinearEstimator
from h2o.grid.grid_search import H2OGridSearch
# Train and validate a cartesian grid of GBMs
glm_grid1 = H2OGridSearch(model=H2OGeneralizedLinearEstimator,
                          grid_id='glm',
                          hyper_params=glm_params1)

In [273]:
glm_grid1.train(X,y, training_frame=df)

glm Grid Build progress: |████████████████████████████████████████████████| 100%


In [274]:
glm_grid1 = glm_grid1.get_grid(sort_by='rmse')
glm_grid1

     tweedie_variance_power     model_ids                 rmse
0                      -1.0  glm_model_11   0.2334520226462017
1                       0.0  glm_model_15   0.2334520226462017
2                       1.0  glm_model_19   0.2334520226462017
3                      -1.0  glm_model_14  0.23345202415806532
4                       0.0  glm_model_18  0.23345202415806532
5                       1.0  glm_model_22  0.23345202415806532
6                       0.0  glm_model_17  0.23345202415914754
7                      -1.0  glm_model_13  0.23345202415914754
8                       1.0  glm_model_21  0.23345202415914754
9                       0.0   glm_model_7  0.23345202416059152
10                      0.0   glm_model_2  0.23345202416059152
11                      1.0   glm_model_3  0.23345202416059152
12                     -1.0   glm_model_6  0.23345202416059152
13                     -1.0   glm_model_1  0.23345202416059152
14                     -1.0   glm_model_4  0.2334520241

**alpha is making changes in the sore, thus this is also not the important hyperparameter**

In [275]:
glm_params1 = { 'tweedie_variance_power': [-1,0,1],
                 'alpha' : [0,0.2,0.7,1],
                'lambda': [0.0013175999248032143,0.09590749037478274, 0.022970099250379815, 0.003416462552479351,0.40044436071981493]
#                 'theta': [888573333419439000,711766579752531000,6743050173023040000]
              }

In [276]:
from h2o.estimators import H2ORandomForestEstimator
from h2o.estimators import H2OGeneralizedLinearEstimator
from h2o.grid.grid_search import H2OGridSearch
# Train and validate a cartesian grid of GBMs
glm_grid1 = H2OGridSearch(model=H2OGeneralizedLinearEstimator,
                          grid_id='glm',
                          hyper_params=glm_params1)

In [277]:
glm_grid1.train(X,y, training_frame=df)

glm Grid Build progress: |████████████████████████████████████████████████| 100%


In [278]:
glm_grid1 = glm_grid1.get_grid(sort_by='rmse')
glm_grid1

      tweedie_variance_power     model_ids                 rmse
0                       -1.0  glm_model_11   0.2334520226462017
1                        0.0  glm_model_15   0.2334520226462017
2                        1.0  glm_model_19   0.2334520226462017
3                        1.0  glm_model_63  0.23345202283341493
4                        0.0  glm_model_43  0.23345202283341493
5                       -1.0  glm_model_23  0.23345202283341493
6                        1.0  glm_model_75   0.2334520238986759
7                       -1.0  glm_model_35   0.2334520238986759
8                        0.0  glm_model_55   0.2334520238986759
9                       -1.0  glm_model_14  0.23345202415806532
10                       0.0  glm_model_18  0.23345202415806532
11                       1.0  glm_model_22  0.23345202415806532
12                       0.0  glm_model_17  0.23345202415914754
13                      -1.0  glm_model_13  0.23345202415914754
14                       1.0  glm_model_

**Lambda is not affecting the score thus we can consider it as imporatant hyperparameter**

In [279]:
glm_params1 = { 'tweedie_variance_power': [-1,0,1],
                 'alpha' : [0,0.2,0.7,1],
                'lambda': [0.0013175999248032143,0.09590749037478274, 0.022970099250379815, 0.003416462552479351,0.40044436071981493],
                'theta': [888573333419439000,711766579752531000,6743050173023040000]
              }

In [280]:
glm_grid1.train(X,y, training_frame=df)

glm Grid Build progress: |████████████████████████████████████████████████| 100%


In [281]:
glm_grid1 = glm_grid1.get_grid(sort_by='rmse')
glm_grid1

      tweedie_variance_power     model_ids                 rmse
0                       -1.0  glm_model_11   0.2334520226462017
1                        0.0  glm_model_15   0.2334520226462017
2                        1.0  glm_model_19   0.2334520226462017
3                        1.0  glm_model_63  0.23345202283341493
4                        0.0  glm_model_43  0.23345202283341493
5                       -1.0  glm_model_23  0.23345202283341493
6                        1.0  glm_model_75   0.2334520238986759
7                       -1.0  glm_model_35   0.2334520238986759
8                        0.0  glm_model_55   0.2334520238986759
9                       -1.0  glm_model_14  0.23345202415806532
10                       0.0  glm_model_18  0.23345202415806532
11                       1.0  glm_model_22  0.23345202415806532
12                       0.0  glm_model_17  0.23345202415914754
13                      -1.0  glm_model_13  0.23345202415914754
14                       1.0  glm_model_

**theta is not affecting the score as compared from the previous score, thus we can not consider it as important hyperparameter**

In [282]:
glm_params1 = { 'tweedie_link_power': [-1,0,1,2],'alpha' : [0,0.2,0.7,1],
                'lambda': [0.0013175999248032143,0.09590749037478274, 0.022970099250379815, 0.003416462552479351,0.40044436071981493],
                'theta': [888573333419439000,711766579752531000,6743050173023040000]
              }

In [283]:
glm_grid1.train(X,y, training_frame=df)

glm Grid Build progress: |████████████████████████████████████████████████| 100%


In [284]:
glm_grid1 = glm_grid1.get_grid(sort_by='rmse')
glm_grid1

      tweedie_variance_power     model_ids                 rmse
0                       -1.0  glm_model_11   0.2334520226462017
1                        0.0  glm_model_15   0.2334520226462017
2                        1.0  glm_model_19   0.2334520226462017
3                        1.0  glm_model_63  0.23345202283341493
4                        0.0  glm_model_43  0.23345202283341493
5                       -1.0  glm_model_23  0.23345202283341493
6                        1.0  glm_model_75   0.2334520238986759
7                       -1.0  glm_model_35   0.2334520238986759
8                        0.0  glm_model_55   0.2334520238986759
9                       -1.0  glm_model_14  0.23345202415806532
10                       0.0  glm_model_18  0.23345202415806532
11                       1.0  glm_model_22  0.23345202415806532
12                       0.0  glm_model_17  0.23345202415914754
13                      -1.0  glm_model_13  0.23345202415914754
14                       1.0  glm_model_

**tweedie_link_power is not affecting the score as compared from the previous score, thus we can not consider it as important hyperparameter**

## Grid search for GBM

In [305]:
from h2o.estimators.gbm import H2OGradientBoostingEstimator
gbm_params1 = {
   'learn_rate': [0.001,0.1,0.5],
#    'col_sample_rate': [0.4,0.7,0.8,1],
#    #'tweedie': [1,1.5,2],
#    'huber_alpha' : [0.8,0.9,1],
#    'ntrees' : [3,4,18,30,31,252,295],
#    'max_depth' : [3,4,5,6,16,17],
#    'min_rows' : [1,5,10,15,30,100],
#    'sample_rate' : [0.5,0.6,0.7,0.8,0.9,1],
#    'col_sample_rate_per_tree' : [0.4,0.7,0.8,1]
}

# Instantiate the grid search model
gbm_grid1 = H2OGridSearch(model=H2OGradientBoostingEstimator,
                         grid_id='gbm_grid1', hyper_params = gbm_params1)

In [306]:
gbm_grid1.train(X, y, training_frame=df)

gbm Grid Build progress: |████████████████████████████████████████████████| 100%


In [307]:
# Get the grid results, sorted by validation AUC
gbm_gridperf1 = gbm_grid1.get_grid(sort_by='rmse')
gbm_gridperf1

       learn_rate            model_ids                 rmse
0             0.5    gbm_grid1_model_3  0.19841074557040178
1             0.5   gbm_grid1_model_12   0.2084625111164261
2             0.5   gbm_grid1_model_62    0.210377133706633
3             0.5   gbm_grid1_model_71  0.21063808501624076
4             0.5   gbm_grid1_model_72  0.21123134328947468
5             0.5   gbm_grid1_model_67  0.21242143972348224
6             0.5  gbm_grid1_model_228  0.21245984853459865
7             0.5  gbm_grid1_model_247  0.21302063338848035
8             0.5   gbm_grid1_model_63   0.2130250405500896
9             0.5  gbm_grid1_model_240  0.21333172947582874
10            0.5  gbm_grid1_model_215  0.21333175316274308
11            0.5  gbm_grid1_model_242  0.21343261025369958
12            0.5  gbm_grid1_model_104  0.21398369042206555
13            0.5  gbm_grid1_model_100  0.21398369042206555
14            0.5  gbm_grid1_model_108  0.21398369042206555
15            0.5  gbm_grid1_model_244  

In [308]:
from h2o.estimators.gbm import H2OGradientBoostingEstimator
gbm_params1 = {
   'learn_rate': [0.001,0.1,0.5],
    'col_sample_rate': [0.4,0.7,0.8,1]
#    #'tweedie': [1,1.5,2],
#    'huber_alpha' : [0.8,0.9,1],
#    'ntrees' : [3,4,18,30,31,252,295],
#    'max_depth' : [3,4,5,6,16,17],
#    'min_rows' : [1,5,10,15,30,100],
#    'sample_rate' : [0.5,0.6,0.7,0.8,0.9,1],
#    'col_sample_rate_per_tree' : [0.4,0.7,0.8,1]
}

# Instantiate the grid search model
gbm_grid1 = H2OGridSearch(model=H2OGradientBoostingEstimator,
                         grid_id='gbm_grid1', hyper_params = gbm_params1)

gbm_grid1.train(X, y, training_frame=df)


gbm Grid Build progress: |████████████████████████████████████████████████| 100%


In [309]:
# Get the grid results, sorted by validation AUC
gbm_gridperf1 = gbm_grid1.get_grid(sort_by='rmse')
gbm_gridperf1

       learn_rate            model_ids                 rmse
0             0.5    gbm_grid1_model_3  0.19841074557040178
1             0.5   gbm_grid1_model_12   0.2084625111164261
2             0.5   gbm_grid1_model_62    0.210377133706633
3             0.5   gbm_grid1_model_71  0.21063808501624076
4             0.5   gbm_grid1_model_72  0.21123134328947468
5             0.5   gbm_grid1_model_67  0.21242143972348224
6             0.5  gbm_grid1_model_228  0.21245984853459865
7             0.5  gbm_grid1_model_247  0.21302063338848035
8             0.5   gbm_grid1_model_63   0.2130250405500896
9             0.5  gbm_grid1_model_240  0.21333172947582874
10            0.5  gbm_grid1_model_215  0.21333175316274308
11            0.5  gbm_grid1_model_242  0.21343261025369958
12            0.5  gbm_grid1_model_104  0.21398369042206555
13            0.5  gbm_grid1_model_100  0.21398369042206555
14            0.5  gbm_grid1_model_108  0.21398369042206555
15            0.5  gbm_grid1_model_244  

**col_sample_rate** is not affecting the score as compared from the previous score, thus we can not consider it as important hyperparameter

In [310]:
from h2o.estimators.gbm import H2OGradientBoostingEstimator
gbm_params1 = {
    'learn_rate': [0.001,0.1,0.5],
    'col_sample_rate': [0.4,0.7,0.8,1],
   'huber_alpha' : [0.8,0.9,1],
#     'quantile_alpha' : [0.5],
#     'sample_rate':[0.5,1],
#     'min_split_improvement': [0.0001],
#    'col_sample_rate_per_tree' : [0.4,0.7,0.8,1]
}

# Instantiate the grid search model
gbm_grid1 = H2OGridSearch(model=H2OGradientBoostingEstimator,
                         grid_id='gbm_grid1', hyper_params = gbm_params1)

gbm_grid1.train(X, y, training_frame=df)


gbm Grid Build progress: |████████████████████████████████████████████████| 100%


In [311]:
gbm_gridperf1 = gbm_grid1.get_grid(sort_by='rmse')
gbm_gridperf1

       learn_rate            model_ids                 rmse
0             0.5    gbm_grid1_model_3  0.19841074557040178
1             0.5   gbm_grid1_model_12   0.2084625111164261
2             0.5   gbm_grid1_model_62    0.210377133706633
3             0.5   gbm_grid1_model_71  0.21063808501624076
4             0.5   gbm_grid1_model_72  0.21123134328947468
5             0.5   gbm_grid1_model_67  0.21242143972348224
6             0.5  gbm_grid1_model_228  0.21245984853459865
7             0.5  gbm_grid1_model_247  0.21302063338848035
8             0.5   gbm_grid1_model_63   0.2130250405500896
9             0.5  gbm_grid1_model_240  0.21333172947582874
10            0.5  gbm_grid1_model_215  0.21333175316274308
11            0.5  gbm_grid1_model_242  0.21343261025369958
12            0.5  gbm_grid1_model_104  0.21398369042206555
13            0.5  gbm_grid1_model_100  0.21398369042206555
14            0.5  gbm_grid1_model_108  0.21398369042206555
15            0.5  gbm_grid1_model_244  

In [312]:
from h2o.estimators.gbm import H2OGradientBoostingEstimator
gbm_params1 = {
    'learn_rate': [0.001,0.1,0.5],
    'col_sample_rate': [0.4,0.7,0.8,1],
       #'tweedie': [1,1.5,2],
   'huber_alpha' : [0.8,0.9,1],
    'quantile_alpha' : [0.5],
#     'sample_rate':[0.5,1],
#     'min_split_improvement': [0.0001],
#    'col_sample_rate_per_tree' : [0.4,0.7,0.8,1]
}


# Instantiate the grid search model
gbm_grid1 = H2OGridSearch(model=H2OGradientBoostingEstimator,
                         grid_id='gbm_grid1', hyper_params = gbm_params1)

gbm_grid1.train(X, y, training_frame=df)


gbm Grid Build progress: |████████████████████████████████████████████████| 100%


In [313]:
gbm_gridperf1 = gbm_grid1.get_grid(sort_by='rmse')
gbm_gridperf1

       learn_rate            model_ids                 rmse
0             0.5    gbm_grid1_model_3  0.19841074557040178
1             0.5   gbm_grid1_model_12   0.2084625111164261
2             0.5   gbm_grid1_model_62    0.210377133706633
3             0.5   gbm_grid1_model_71  0.21063808501624076
4             0.5   gbm_grid1_model_72  0.21123134328947468
5             0.5   gbm_grid1_model_67  0.21242143972348224
6             0.5  gbm_grid1_model_228  0.21245984853459865
7             0.5  gbm_grid1_model_247  0.21302063338848035
8             0.5   gbm_grid1_model_63   0.2130250405500896
9             0.5  gbm_grid1_model_240  0.21333172947582874
10            0.5  gbm_grid1_model_215  0.21333175316274308
11            0.5  gbm_grid1_model_242  0.21343261025369958
12            0.5  gbm_grid1_model_104  0.21398369042206555
13            0.5  gbm_grid1_model_100  0.21398369042206555
14            0.5  gbm_grid1_model_108  0.21398369042206555
15            0.5  gbm_grid1_model_244  

In [314]:
gbm_gridperf1 = gbm_grid1.get_grid(sort_by='rmse')
gbm_gridperf1

       learn_rate            model_ids                 rmse
0             0.5    gbm_grid1_model_3  0.19841074557040178
1             0.5   gbm_grid1_model_12   0.2084625111164261
2             0.5   gbm_grid1_model_62    0.210377133706633
3             0.5   gbm_grid1_model_71  0.21063808501624076
4             0.5   gbm_grid1_model_72  0.21123134328947468
5             0.5   gbm_grid1_model_67  0.21242143972348224
6             0.5  gbm_grid1_model_228  0.21245984853459865
7             0.5  gbm_grid1_model_247  0.21302063338848035
8             0.5   gbm_grid1_model_63   0.2130250405500896
9             0.5  gbm_grid1_model_240  0.21333172947582874
10            0.5  gbm_grid1_model_215  0.21333175316274308
11            0.5  gbm_grid1_model_242  0.21343261025369958
12            0.5  gbm_grid1_model_104  0.21398369042206555
13            0.5  gbm_grid1_model_100  0.21398369042206555
14            0.5  gbm_grid1_model_108  0.21398369042206555
15            0.5  gbm_grid1_model_244  

In [315]:
from h2o.estimators.gbm import H2OGradientBoostingEstimator
gbm_params1 = {
    'learn_rate': [0.001,0.1,0.5],
    'col_sample_rate': [0.4,0.7,0.8,1],
       #'tweedie': [1,1.5,2],
   'huber_alpha' : [0.8,0.9,1],
    'quantile_alpha' : [0.5],
    'sample_rate':[0.5,1],
     'min_split_improvement': [0.0001],
#    'col_sample_rate_per_tree' : [0.4,0.7,0.8,1]
}

# Instantiate the grid search model
gbm_grid1 = H2OGridSearch(model=H2OGradientBoostingEstimator,
                         grid_id='gbm_grid1', hyper_params = gbm_params1)

gbm_grid1.train(X, y, training_frame=df)



gbm Grid Build progress: |████████████████████████████████████████████████| 100%


In [316]:
gbm_gridperf1 = gbm_grid1.get_grid(sort_by='rmse')
gbm_gridperf1

       learn_rate            model_ids                 rmse
0             0.5    gbm_grid1_model_3  0.19841074557040178
1             0.5   gbm_grid1_model_12   0.2084625111164261
2             0.5   gbm_grid1_model_62    0.210377133706633
3             0.5   gbm_grid1_model_71  0.21063808501624076
4             0.5   gbm_grid1_model_72  0.21123134328947468
5             0.5   gbm_grid1_model_67  0.21242143972348224
6             0.5  gbm_grid1_model_228  0.21245984853459865
7             0.5  gbm_grid1_model_247  0.21302063338848035
8             0.5   gbm_grid1_model_63   0.2130250405500896
9             0.5  gbm_grid1_model_240  0.21333172947582874
10            0.5  gbm_grid1_model_215  0.21333175316274308
11            0.5  gbm_grid1_model_242  0.21343261025369958
12            0.5  gbm_grid1_model_104  0.21398369042206555
13            0.5  gbm_grid1_model_100  0.21398369042206555
14            0.5  gbm_grid1_model_108  0.21398369042206555
15            0.5  gbm_grid1_model_244  

In [317]:
from h2o.estimators.gbm import H2OGradientBoostingEstimator
gbm_params1 = {
    'learn_rate': [0.001,0.1,0.5],
    'col_sample_rate': [0.4,0.7,0.8,1],
       #'tweedie': [1,1.5,2],
   'huber_alpha' : [0.8,0.9,1],
    'quantile_alpha' : [0.5],
    'sample_rate':[0.5,1],
     'min_split_improvement': [0.0001],
   'col_sample_rate_per_tree' : [0.4,0.7,0.8,1]
}

# Instantiate the grid search model
gbm_grid1 = H2OGridSearch(model=H2OGradientBoostingEstimator,
                         grid_id='gbm_grid1', hyper_params = gbm_params1)

gbm_grid1.train(X, y, training_frame=df)



gbm Grid Build progress: |████████████████████████████████████████████████| 100%


In [318]:
gbm_gridperf1 = gbm_grid1.get_grid(sort_by='rmse')
gbm_gridperf1

       learn_rate            model_ids                 rmse
0             0.5    gbm_grid1_model_3  0.19841074557040178
1             0.5   gbm_grid1_model_12   0.2084625111164261
2             0.5   gbm_grid1_model_62    0.210377133706633
3             0.5   gbm_grid1_model_71  0.21063808501624076
4             0.5   gbm_grid1_model_72  0.21123134328947468
5             0.5   gbm_grid1_model_67  0.21242143972348224
6             0.5  gbm_grid1_model_228  0.21245984853459865
7             0.5  gbm_grid1_model_247  0.21302063338848035
8             0.5   gbm_grid1_model_63   0.2130250405500896
9             0.5  gbm_grid1_model_240  0.21333172947582874
10            0.5  gbm_grid1_model_215  0.21333175316274308
11            0.5  gbm_grid1_model_242  0.21343261025369958
12            0.5  gbm_grid1_model_104  0.21398369042206555
13            0.5  gbm_grid1_model_100  0.21398369042206555
14            0.5  gbm_grid1_model_108  0.21398369042206555
15            0.5  gbm_grid1_model_244  

In [319]:
from h2o.estimators.gbm import H2OGradientBoostingEstimator
gbm_params1 = {
    
    'col_sample_rate': [0.4,0.7,0.8,1],
       #'tweedie': [1,1.5,2],
   'huber_alpha' : [0.8,0.9,1],
    'quantile_alpha' : [0.5],
    'sample_rate':[0.5,1],
     'min_split_improvement': [0.0001],
   'col_sample_rate_per_tree' : [0.4,0.7,0.8,1]
}

# Instantiate the grid search model
gbm_grid1 = H2OGridSearch(model=H2OGradientBoostingEstimator,
                         grid_id='gbm_grid1', hyper_params = gbm_params1)

gbm_grid1.train(X, y, training_frame=df)



gbm Grid Build progress: |████████████████████████████████████████████████| 100%


In [320]:
gbm_gridperf1 = gbm_grid1.get_grid(sort_by='rmse')
gbm_gridperf1

       learn_rate            model_ids                 rmse
0             0.5    gbm_grid1_model_3  0.19841074557040178
1             0.5   gbm_grid1_model_12   0.2084625111164261
2             0.5   gbm_grid1_model_62    0.210377133706633
3             0.5   gbm_grid1_model_71  0.21063808501624076
4             0.5   gbm_grid1_model_72  0.21123134328947468
5             0.5   gbm_grid1_model_67  0.21242143972348224
6             0.5  gbm_grid1_model_228  0.21245984853459865
7             0.5  gbm_grid1_model_247  0.21302063338848035
8             0.5   gbm_grid1_model_63   0.2130250405500896
9             0.5  gbm_grid1_model_240  0.21333172947582874
10            0.5  gbm_grid1_model_215  0.21333175316274308
11            0.5  gbm_grid1_model_242  0.21343261025369958
12            0.5  gbm_grid1_model_104  0.21398369042206555
13            0.5  gbm_grid1_model_100  0.21398369042206555
14            0.5  gbm_grid1_model_108  0.21398369042206555
15            0.5  gbm_grid1_model_244  

In [323]:
h2o.cluster().shutdown()

H2O session _sid_a392 closed.
